### ISSUE TITLE
DHCP won't function /w host or bridge mode.

### ISSUE THREAD

**Issue by Dulanic** [pi-hole/docker-pi-hole#331]
This is a... 

- [ ] Request for a new or modified feature
- [x]  Issue trying to run the docker image
- [ ] Issue trying to build / test / develop the docker image

## Description
DHCP will not work /w either bridge or host mode.

## Expected Behavior
Pi.Hole would act as my DHCP server.

## Actual Behavior
The container is not announcing or acting as a available DHCP server.
```
C:\Users\dulanic>ipconfig /renew

Windows IP Configuration

An error occurred while renewing interface Ethernet : unable to contact your DHCP server. Request has timed out.

```
## Possible Fix
Have not found one.

## Steps to Reproduce and debugging done
```
#!/bin/bash
# Lookups may not work for VPN / tun0
IP_LOOKUP="$(ip route get 8.8.8.8 | awk '{ print $NF; exit }')"
IPv6_LOOKUP="$(ip -6 route get 2001:4860:4860::8888 | awk '{for(i=1;i<=NF;i++) i                                              f ($i=="src") print $(i+1)}')"

# Just hard code these to your docker server's LAN IP if lookups aren't working
IP="${IP:-$IP_LOOKUP}"  # use $IP, if set, otherwise IP_LOOKUP
IPv6="${IPv6:-$IPv6_LOOKUP}"  # use $IPv6, if set, otherwise IP_LOOKUP

# Default of directory you run this from, update to where ever.
DOCKER_CONFIGS="$(pwd)"

echo "### Make sure your IPs are correct, hard code ServerIP ENV VARs if necessa                                              ry\nIP: ${IP}\nIPv6: ${IPv6}"

# Default ports + daemonized docker container
docker run -d \
    --name pihole \
    --net=host \
    -p 53:53/tcp -p 53:53/udp \
    --dns 127.0.0.1 --dns 1.1.1.1 \
    -p 67:67/udp \
    -p 80:80 \
    -p 443:443 \
    -v "${DOCKER_CONFIGS}/pihole/:/etc/pihole/" \
    -v "${DOCKER_CONFIGS}/dnsmasq.d/:/etc/dnsmasq.d/" \
    -e ServerIP="${IP}" \
    -e TZ=America/Chicago \
    -e DNS1=127.0.0.1 \
    -e DNS2=1.1.1.1 \
    -e WEBPASSWORD= \
    --restart=unless-stopped \
    pihole/pihole:v4.0_armhf

echo -n "Your password for https://${IP}/admin/ is "
docker logs pihole 2> /dev/null | grep 'password:'
```

## Debug steps I have tried

- [x] I have tried destroying my container instance, pulling the newest image version, and re-creating a new container
- [x] I have tried running the nearly stock `docker run` example in the readme (removing any customizations I added)
- [x] I have tried running without my volume data mounts to eliminate volumes as the cause
- [x] I have searched this repository for existing issues and pull requests that look similar 

## Context and extra information
IPTables info:
```

pi@raspberrypi:~ $ sudo iptables --list
Chain INPUT (policy ACCEPT)
target     prot opt source               destination

Chain FORWARD (policy DROP)
target     prot opt source               destination
DOCKER-USER  all  --  anywhere             anywhere
DOCKER-ISOLATION-STAGE-1  all  --  anywhere             anywhere
ACCEPT     all  --  anywhere             anywhere             ctstate RELATED,ESTABLISHED
DOCKER     all  --  anywhere             anywhere
ACCEPT     all  --  anywhere             anywhere
ACCEPT     all  --  anywhere             anywhere

Chain OUTPUT (policy ACCEPT)
target     prot opt source               destination

Chain DOCKER (1 references)
target     prot opt source               destination
ACCEPT     tcp  --  anywhere             172.17.0.2           tcp dpt:9000

Chain DOCKER-ISOLATION-STAGE-1 (1 references)
target     prot opt source               destination
DOCKER-ISOLATION-STAGE-2  all  --  anywhere             anywhere
RETURN     all  --  anywhere             anywhere

Chain DOCKER-ISOLATION-STAGE-2 (1 references)
target     prot opt source               destination
DROP       all  --  anywhere             anywhere
RETURN     all  --  anywhere             anywhere

Chain DOCKER-USER (1 references)
target     prot opt source               destination
RETURN     all  --  anywhere             anywhere

```

## Your Environment 
Raspberry Pi 3
```
Client:
 Version:           18.06.1-ce
 API version:       1.38
 Go version:        go1.10.3
 Git commit:        e68fc7a
 Built:             Tue Aug 21 17:30:52 2018
 OS/Arch:           linux/arm
 Experimental:      false

Server:
 Engine:
  Version:          18.06.1-ce
  API version:      1.38 (minimum version 1.12)
  Go version:       go1.10.3
  Git commit:       e68fc7a
  Built:            Tue Aug 21 17:26:37 2018
  OS/Arch:          linux/arm
  Experimental:     false
```

**Comment by Dulanic** [comment]
OK here is the results from those:
```

pi@raspberrypi:~ $ sudo ps -efH | grep 'docker-proxy'
sudo: unable to resolve host raspberrypi.home
root       785   498  0 Sep15 ?        00:00:00     /usr/bin/docker-proxy -proto tcp -host-ip 0.0.0.0 -host-port 9000 -container-ip 172.17.0.2 -container-port 9000
root     23297   498  0 16:42 ?        00:00:00     /usr/bin/docker-proxy -proto tcp -host-ip 0.0.0.0 -host-port 443 -container-ip 172.17.0.4 -container-port 443
root     23311   498  0 16:42 ?        00:00:00     /usr/bin/docker-proxy -proto tcp -host-ip 0.0.0.0 -host-port 80 -container-ip 172.17.0.4 -container-port 80
root     23324   498  0 16:42 ?        00:00:00     /usr/bin/docker-proxy -proto udp -host-ip 0.0.0.0 -host-port 67 -container-ip 172.17.0.4 -container-port 67
root     23339   498  0 16:42 ?        00:00:00     /usr/bin/docker-proxy -proto tcp -host-ip 0.0.0.0 -host-port 53 -container-ip 172.17.0.4 -container-port 53
root     23352   498  0 16:42 ?        00:00:00     /usr/bin/docker-proxy -proto udp -host-ip 0.0.0.0 -host-port 53 -container-ip 172.17.0.4 -container-port 53
pi       24208 22815  0 16:43 pts/0    00:00:00           grep --color=auto docker-proxy

```
```

pi@raspberrypi:~ $ sudo netstat -nap | grep -E ':(53|67|80|443) '
sudo: unable to resolve host raspberrypi.home
tcp6       0      0 :::80                   :::*                    LISTEN      23311/docker-proxy
tcp6       0      0 :::53                   :::*                    LISTEN      23339/docker-proxy
tcp6       0      0 :::443                  :::*                    LISTEN      23297/docker-proxy
udp        0      0 172.17.0.1:36940        172.17.0.4:53           ESTABLISHED 23352/docker-proxy
udp        0      0 172.17.0.1:54941        172.17.0.4:67           ESTABLISHED 23324/docker-proxy
udp        0      0 172.17.0.1:52565        172.17.0.4:53           ESTABLISHED 23352/docker-proxy
udp6       0      0 :::53                   :::*                                23352/docker-proxy
udp6       0      0 :::67                   :::*                                23324/docker-proxy

```
```

pi@raspberrypi:~ $ nc -vz -u localhost 67 || echo failed to connect
Connection to localhost 67 port [udp/bootps] succeeded!
```

Yet it will won't work. My phone says can't get a IP address, my windows PC won't either.

![image](https://user-images.githubusercontent.com/2948206/45637649-a14bc100-ba70-11e8-82dc-43e7736f3074.png)

**Comment by Dulanic** [comment]
Wireshark traffic, start is when my router DHCP is on, then after I turn it off. The pi.hole DHCP is enabled the whole time.

![image](https://user-images.githubusercontent.com/2948206/45639474-b840e200-ba75-11e8-9463-485e000b4b5c.png)

**Comment by Dulanic** [comment]
Poweshell shows the docker container is actively refusing connection. So it seems to be an issue /w the source of the broadcast? Obviously localhost it works fine.

![image](https://user-images.githubusercontent.com/2948206/45640589-b75d7f80-ba78-11e8-909a-30be5869d201.png)

**Comment by diginc** [comment]
I think you may need to configure your router DHCP Relay as the IP of your docker host: 

https://help.ubnt.com/hc/en-us/articles/204962294-EdgeRouter-DHCP-Relay
https://en.wikipedia.org/wiki/Dynamic_Host_Configuration_Protocol#DHCP_relaying

**Comment by diginc** [comment]
Sorry just saw you mentioned it didn't work with --net=host network too...hmm that may shoot a hole in my theory.  Could you try the relay with the bridged / default network mode?

**Comment by diginc** [comment]
Just stated on x86 and arm, regardless of the --net setting you have to configure a DHCP relay on your router to point it at the docker pi-hole.  I'll get the documentation updated accordingly to make this more apparent.

**Comment by DerFetzer** [comment]
With `net=host` I do not need a relay on my router. With this the docker container is on the same subnet as all other devices in the network.

**Comment by Dulanic** [comment]
Sorry for the delay in testing, yeah I didn't get to. Thanks for looking into it. I don';t get why that happens /w net=host since it isn't a separate subnet though. I even went into the container and that SHOULD work.....

```

1: lo: <LOOPBACK,UP,LOWER_UP> mtu 65536 qdisc noqueue state UNKNOWN group default qlen 1000
    link/loopback 00:00:00:00:00:00 brd 00:00:00:00:00:00
    inet 127.0.0.1/8 scope host lo
       valid_lft forever preferred_lft forever
    inet6 ::1/128 scope host
       valid_lft forever preferred_lft forever
2: eth0: <BROADCAST,MULTICAST,UP,LOWER_UP> mtu 1500 qdisc pfifo_fast state UP group default qlen 1000
    link/ether b8:27:eb:d0:b3:0d brd ff:ff:ff:ff:ff:ff
    inet 192.168.2.250/24 brd 192.168.2.255 scope global eth0
       valid_lft forever preferred_lft forever
    inet6 fe80::8515:e5e0:9ac2:52fa/64 scope link
       valid_lft forever preferred_lft forever
3: wlan0: <NO-CARRIER,BROADCAST,MULTICAST,UP> mtu 1500 qdisc pfifo_fast state DOWN group default qlen 1000
    link/ether b8:27:eb:85:e6:58 brd ff:ff:ff:ff:ff:ff
4: docker0: <BROADCAST,MULTICAST,UP,LOWER_UP> mtu 1500 qdisc noqueue state UP group default
    link/ether 02:42:86:b0:b5:67 brd ff:ff:ff:ff:ff:ff
    inet 172.17.0.1/16 brd 172.17.255.255 scope global docker0
       valid_lft forever preferred_lft forever
    inet 169.254.82.227/16 brd 169.254.255.255 scope global docker0
       valid_lft forever preferred_lft forever
    inet6 fe80::e49a:a305:58c5:c2b/64 scope link
       valid_lft forever preferred_lft forever
6: vethde3f342@if5: <BROADCAST,MULTICAST,UP,LOWER_UP> mtu 1500 qdisc noqueue master docker0 state UP group default
    link/ether a6:7d:45:ba:40:23 brd ff:ff:ff:ff:ff:ff link-netnsid 0
    inet 169.254.226.89/16 brd 169.254.255.255 scope global vethde3f342
       valid_lft forever preferred_lft forever
    inet6 fe80::a47d:45ff:feba:4023/64 scope link
       valid_lft forever preferred_lft forever
16: veth5171a03@if15: <BROADCAST,MULTICAST,UP,LOWER_UP> mtu 1500 qdisc noqueue master docker0 state UP groupdefault
    link/ether e6:c8:65:68:04:62 brd ff:ff:ff:ff:ff:ff link-netnsid 1
    inet 169.254.135.69/16 brd 169.254.255.255 scope global veth5171a03
       valid_lft forever preferred_lft forever
    inet6 fe80::e4c8:65ff:fe68:462/64 scope link
       valid_lft forever preferred_lft forever
```

**Comment by diginc** [comment]
I noticed you had missed `--cap-add=NET_ADMIN` from your docker run, any chance you added that.  I think I missed it in my test too since I copied your command exactly.

**Comment by Dulanic** [comment]
I doubt anything changed because I use the script and I literally just changed the net=host. I even left the -p assignments not that it matters since it doesnt use them for host.
```

pi@raspberrypi:~ $ cat docker*
#!/bin/bash
# Lookups may not work for VPN / tun0
IP_LOOKUP="$(ip route get 8.8.8.8 | awk '{ print $NF; exit }')"
IPv6_LOOKUP="$(ip -6 route get 2001:4860:4860::8888 | awk '{for(i=1;i<=NF;i++) if ($i=="src") print $(i+1)}')"

# Just hard code these to your docker server's LAN IP if lookups aren't working
IP="${IP:-$IP_LOOKUP}"  # use $IP, if set, otherwise IP_LOOKUP
IPv6="${IPv6:-$IPv6_LOOKUP}"  # use $IPv6, if set, otherwise IP_LOOKUP

# Default of directory you run this from, update to where ever.
DOCKER_CONFIGS="$(pwd)"

echo "### Make sure your IPs are correct, hard code ServerIP ENV VARs if necessary\nIP: ${IP}\nIPv6: ${IPv6}"

# Default ports + daemonized docker container
docker run -d \
    --name pihole \
    --net=host \
    -p 53:53/tcp -p 53:53/udp \
    --dns 127.0.0.1 --dns 1.1.1.1 \
    -p 67:67/udp \
    -p 80:80 \
    -p 443:443 \
    -v "${DOCKER_CONFIGS}/pihole/:/etc/pihole/" \
    -v "${DOCKER_CONFIGS}/dnsmasq.d/:/etc/dnsmasq.d/" \
    -e ServerIP="${IP}" \
    -e TZ=America/Chicago \
    -e DNS1=127.0.0.1 \
    -e DNS2=1.1.1.1 \
    -e WEBPASSWORD= \
    --restart=unless-stopped \
    pihole/pihole:v4.0_armhf

echo -n "Your password for https://${IP}/admin/ is "
docker logs pihole 2> /dev/null | grep 'password:'
```

**Comment by diginc** [comment]
Interesting, I tried getting an IP on various clients for a solid 15 minutes with net=host set.  Then as soon as I put in a DHCP relay pointed to the RBPi docker pi-hole it instantly started working.  I must've been missing some other piece of the networking puzzle...in theory net host should work without a relay, I totally agree there.

For anyone one else reading this trying to debug DHCP, here is what I was using from a linux client to try to renew DHCP as I tweaked the Pi-hole server / router relay settings :  `sudo dhclient -r && sudo dhclient -v` (release lease, renew lease with verbosity)

There's also some other helpful commands here I didn't try last night but I may check out dhcpdump in the future : https://www.cyberciti.biz/faq/linux-unix-dhcpdump-monitor-dhcp-traffic/
